crates/autopilot/src/limit_orders/quoter.rs

```diff
@@ -33,40 +31,43 @@ pub struct LimitOrderQuoter {
 
 impl LimitOrderQuoter {
     pub fn spawn(self) {
-        tokio::spawn(async move {
-            loop {
-                if let Err(err) = self.update().await {
-                    tracing::error!(?err, "failed to update limit order surplus");
-                }
-                tokio::time::sleep(self.loop_delay).await;
-            }
-        });
+        tokio::spawn(async move { self.background_task().await });
     }
 
-    async fn update(&self) -> Result<()> {
+    async fn background_task(&self) -> ! {
         loop {
-            let orders = self
-                .database
-                .limit_orders_with_outdated_fees(self.limit_order_age)
-                .await?;
-            if orders.is_empty() {
-                break;
-            }
-            for chunk in orders.chunks(10) {
-                join_all(chunk.iter().map(|order| {
-                    async move {
-                        let quote = self.get_quote(order).await;
-                        self.update_fee(&order.metadata.uid, &quote).await;
-                    }
-                    .instrument(tracing::debug_span!(
-                        "surplus_fee",
-                        order =% order.metadata.uid
-                    ))
-                }))
-                .await;
-            }
+            let sleep = match self.update().await {
+                Ok(0) => Duration::from_secs_f32(10.),
+                Ok(_) => Duration::from_secs_f32(0.),
+                Err(err) => {
+                    tracing::error!(?err, "limit order quoter update error");
+                    Duration::from_secs_f32(1.)
+                }
+            };
+            tokio::time::sleep(sleep).await;
         }
-        Ok(())
+    }
+
+    /// Returns how many orders were updated.
+    async fn update(&self) -> Result<usize> {
+        const PARALLELISM: usize = 5;
+        let orders = self
+            .database
+            .limit_orders_with_outdated_fees(self.limit_order_age, PARALLELISM)
+            .await?;
+        futures::stream::iter(&orders)
+            .for_each_concurrent(PARALLELISM, |order| {
```

Comment:
Is there any advantage of using a stream compared to `join_all` here?
Not that it's a problem I just think usually streams are a bit harder to work with and to understand (I might be alone on this, though).

Reply:
I felt the stream semantics were clearer than join_all. for_each_concurrente feels clear to me and it shows that I don't care about the individual futures because they never leave the function.

Reply:
I also personally prefer `for_each_concurrent()` over `for _ in chunked() { join_all }`.